File: Question/22_Generate_Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        dfs(ans, 0, 0, n, "");
        return ans;
    }
    void dfs(vector<string>& ans, int left, int right, int n, string cur) {
        if((left == right) && left == n) {
            ans.push_back(cur);
        }
        if(left < n)
            dfs(ans, left + 1, right, n, cur + '(');
        if(right < left)
            dfs(ans, left, right + 1, n, cur + ')');
    }
};

File: Question_Go/22_Generate_Parentheses.go
func generateParenthesis(n int) []string {
	if n == 0 {
		return nil
	}

	ans := []string{}
	generate(n, n, "", &ans)
	return ans
}

func generate(left int, right int, cur string, ans *[]string) {
	if left == 0 && right == 0 {
		*ans = append(*ans, cur)
		return
	}
	if left < 0 || right < 0 {
		return
	}
	if left > right {
		return
	}

	//left and right
	generate(left-1, right, cur+"(", ans)
	generate(left, right-1, cur+")", ans)
}